crates/test_probe/src/main.rs

```diff
@@ -37,41 +46,66 @@ fn main() -> Result<()> {
         Some(address) => Client::new(address),
         None => Client::try_default(),
     }?;
-    let event = Event::TestTick;
 
-    let send = || {
-        if let Err(err) = client.send(&event) {
+    let send = |event: &Event| {
+        if let Err(err) = client.send(event) {
             eprintln!("Could not send event {:?}", err)
         };
     };
 
     match args.mode {
         Mode::Constant { frequency } => constant_frequency(frequency, send),
         Mode::Rollercoaster => rollercoaster(send),
+        Mode::Burst { burst_period_ms, events_per_burst } => {
+            burst(Duration::from_millis(burst_period_ms), events_per_burst, send)
+        },
     }
 }
 
-fn constant_frequency(frequency: f64, send_fn: impl Fn()) -> ! {
+fn constant_frequency(frequency: f64, send: impl Fn(&Event)) -> ! {
+    let event = Event::TestTick;
+
     // Try to prevent drifting away from the given frequency by dynamically computing the sleep duration for each cycle
     let start = Instant::now();
     for deadline in (0..).map(|i| start + Duration::from_secs_f64(i as f64 / frequency)) {
-        send_fn();
+        send(&event);
         sleep(deadline.saturating_duration_since(Instant::now()));
     }
 
     unreachable!()
 }
 
-fn rollercoaster(send_fn: impl Fn()) -> ! {
+fn rollercoaster(send: impl Fn(&Event)) -> ! {
+    let event = Event::TestTick;
+
     // u64 taken by `from_millis` doesn't implement DoubleEndedIterator needed by `rev`
     // Use u32 explicitly and convert to u64.
     let slowdown = (5u32..200).step_by(5);
     let speedup = slowdown.clone().rev();
 
     for delay_ms in speedup.chain(slowdown).cycle() {
-        send_fn();
+        send(&event);
         sleep(Duration::from_millis(delay_ms.into()));
     }
 
     unreachable!()
 }
+
+fn burst(burst_period: Duration, events_per_burst: u32, send: impl Fn(&Event)) -> ! {
+    let event_period = burst_period / events_per_burst;
+    let start_unix =
+        SystemTime::now().duration_since(UNIX_EPOCH).expect("Couldn't get unix timestamp");
+    let mut event_generator =
+        (0..).map(|n| Event::TimedTestTick { ts: start_unix + n * event_period });
+
+    let start = Instant::now();
+    for deadline in (0..).map(|i| start + i * burst_period) {
+        event_generator.by_ref().take(events_per_burst as usize).for_each(|event| {
+            send(&event);
+        });
+
+        sleep(deadline.saturating_duration_since(Instant::now()));
+    }
```

Comment:
As the code stands, all the events that are sent out at time `T` have timestamps `> T`. That's the opposite of what a typical real-world aggregating probe would produce. Such probe would be collecting events and then it would send them out with timestamps set in past.

I tried to mimic this by setting
```rust
let start_unix = SystemTime::now()
    .duration_since(UNIX_EPOCH)            👇
    .expect("Couldn't get unix timestamp") - burst_period;
```

What I'd expect then, is that once `burst_period` exceeds `AudioOutput`'s `play_delay` we'll start getting "too early plays" in every burst.

But that wasn't the case! Instead, only the first burst is "too early" and all subsequent bursts are played smoothly. I can't figure out how that's possible. 🤔

Reply:
:thinking:

Reply:
Aha! It was caused by a off-by-one bug in the burst for-loop. Fixed in https://github.com/tonarino/acoustic_profiler/pull/44. .)